### PR TITLE
Protect txnReq.newEvents with mutex

### LINKED ISSUE #1584: Concurrent map write in federation API /send
```
time="2020-11-16T18:02:51.637854796Z" level=warning msg="Outgoing request failed" func="DoHTTPRequest\n\t" file=" [github.com/matrix-org/gomatrixserverlib@v0.0.0-20201020162226-22169fe9cda7/client.go:417]" context=missing error="Get \"matrix://rdmnn.de/_matrix/federation/v1/event/$OANfOGD46uGt3CHAlnsfC3yeGdM9QwnfAriRLUXvHGU\": context deadline exceeded" out.req.ID=wg0mDg1dfjJV out.req.method=GET out.req.uri="matrix://rdmnn.de/_matrix/federation/v1/event/$OANfOGD46uGt3CHAlnsfC3yeGdM9QwnfAriRLUXvHGU"
time="2020-11-16T18:02:51.637951305Z" level=warning msg="Failed to get missing /event for event ID" func="lookupEvent\n\t" file=" [github.com/matrix-org/dendrite/federationapi/routing/send.go:1127]" context=missing error="Get \"matrix://rdmnn.de/_matrix/federation/v1/event/$OANfOGD46uGt3CHAlnsfC3yeGdM9QwnfAriRLUXvHGU\": context deadline exceeded" event_id="$OANfOGD46uGt3CHAlnsfC3yeGdM9QwnfAriRLUXvHGU"
time="2020-11-16T18:02:51.645113165Z" level=warning msg="Error sending request to https://matrix.koyu.space:8448/_matrix/federation/v1/event/$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso: context deadline exceeded" func="RoundTrip\n\t" file=" [github.com/matrix-org/gomatrixserverlib@v0.0.0-20201020162226-22169fe9cda7/client.go:146]" context=missing out.req.ID=6UPAN2HFOmJ5 out.req.method=GET out.req.uri="matrix://matrix.koyu.space/_matrix/federation/v1/event/$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso"
time="2020-11-16T18:02:51.645192083Z" level=warning msg="Outgoing request failed" func="DoHTTPRequest\n\t" file=" [github.com/matrix-org/gomatrixserverlib@v0.0.0-20201020162226-22169fe9cda7/client.go:417]" context=missing error="Get \"matrix://matrix.koyu.space/_matrix/federation/v1/event/$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso\": context deadline exceeded" out.req.ID=6UPAN2HFOmJ5 out.req.method=GET out.req.uri="matrix://matrix.koyu.space/_matrix/federation/v1/event/$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso"
time="2020-11-16T18:02:51.645283237Z" level=warning msg="Failed to get missing /event for event ID" func="lookupEvent\n\t" file=" [github.com/matrix-org/dendrite/federationapi/routing/send.go:1127]" context=missing error="Get \"matrix://matrix.koyu.space/_matrix/federation/v1/event/$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso\": context deadline exceeded" event_id="$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso"
time="2020-11-16T18:02:51.645339731Z" level=warning msg="Failed to get missing /event for event ID from 5 server(s)" func="lookupEvent\n\t" file=" [github.com/matrix-org/dendrite/federationapi/routing/send.go:1139]" context=missing event_id="$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso"
time="2020-11-16T18:02:51.645370561Z" level=info msg="Failed to fetch missing event" func="func1\n\t" file=" [github.com/matrix-org/dendrite/federationapi/routing/send.go:1055]" context=missing event_id="$O2hQiD1r1U1-FZo0i3jUUPkS6wqx49CjNuH8QlKGeso" room_id="!OGEhHVWSdvArJzumhm:matrix.org"
fatal error: concurrent map writes

goroutine 2198744 [running]:
runtime.throw(0x110948d, 0x15)
	runtime/panic.go:1116 +0x72 fp=0xc07f075150 sp=0xc07f075120 pc=0x43ff52
runtime.mapassign_faststr(0xf9d260, 0xc03f49af60, 0xc07e1dff80, 0x2c, 0x1)
	runtime/map_faststr.go:211 +0x3f1 fp=0xc07f0751b8 sp=0xc07f075150 pc=0x41d0d1
github.com/matrix-org/dendrite/federationapi/routing.(*txnReq).lookupEvent(0xc01240e000, 0x127a8c0, 0xc05fbe21e0, 0x1816de8, 0x1, 0xc03815d680, 0x2c, 0x0, 0xc07f0f0000, 0x5, ...)
	github.com/matrix-org/dendrite/federationapi/routing/send.go:1147 +0x97d fp=0xc07f075d60 sp=0xc07f0751b8 pc=0xc93d3d
github.com/matrix-org/dendrite/federationapi/routing.(*txnReq).lookupMissingStateViaStateIDs.func1(0xc03815d680, 0x2c)
	github.com/matrix-org/dendrite/federationapi/routing/send.go:1045 +0xdf fp=0xc07f075f80 sp=0xc07f075d60 pc=0xc9c5bf
github.com/matrix-org/dendrite/federationapi/routing.(*txnReq).lookupMissingStateViaStateIDs.func2(0xc07ee4bce0)
	github.com/matrix-org/dendrite/federationapi/routing/send.go:1067 +0x82 fp=0xc07f075fd8 sp=0xc07f075f80 pc=0xc9ca02
runtime.goexit()
	runtime/asm_amd64.s:1374 +0x1 fp=0xc07f075fe0 sp=0xc07f075fd8 pc=0x477a41
created by github.com/matrix-org/dendrite/federationapi/routing.(*txnReq).lookupMissingStateViaStateIDs
	github.com/matrix-org/dendrite/federationapi/routing/send.go:1073 +0xdfe
```

### PR DESCRIPTION
**Pull request by neilalexander**
This should fix #1584.

### CODE REVIEW

**Comment by Kegsay** [comment]
Where are these concurrent writes coming from? `txnReq` is scoped per `/transaction`. Is this because we goroutine-ised the processing of txns per-room but are sharing `txnReq` between them?

**Comment by neilalexander** [comment]
I believe it's because `lookupMissingStateViaStateIDs` starts worker goroutines.

File: federationapi/routing/send.go [hunk 1]
```diff
@@ -1034,11 +1050,6 @@ func (t *txnReq) lookupMissingStateViaStateIDs(ctx context.Context, roomID, even
 	var fetchgroup sync.WaitGroup
 	fetchgroup.Add(concurrentRequests)
 
-	// This is the only place where we'll write to t.haveEvents from
-	// multiple goroutines, and everywhere else is blocked on this
-	// synchronous function anyway.
-	var haveEventsMutex sync.Mutex
```
Comment (Kegsay) [inline federationapi/routing/send.go:1040]: Why wasn't this sufficient then?

File: federationapi/routing/send.go [hunk 2]
```diff
@@ -1144,6 +1158,8 @@ func (t *txnReq) lookupEvent(ctx context.Context, roomVersion gomatrixserverlib.
 		return nil, verifySigError{event.EventID(), err}
 	}
 	h := event.Headered(roomVersion)
+	t.newEventsMutex.Lock()
 	t.newEvents[h.EventID()] = true
```
Comment (Kegsay) [inline federationapi/routing/send.go:1162]: It seems like this was the actual cause of #1584 - as workers will call `lookupEvent` and we don't guard it. Why are we adding unnecessary mutexes - `haveEventsMutex`? Why are we using RWMutexes when really we just do a ton of writes before reading again?

**Review by Kegsay — APPROVED** [review]

**Pull request merged**